### PR TITLE
[9.x] Eloquent Strict Loading

### PR DESCRIPTION
**Pull request by themsaid**
Update: Moved to the 8.x branch: https://github.com/laravel/framework/pull/37363

---

This PR introduces a `strict_load` database connection configuration that when set to `true` throws an exception on lazy loading eloquent relations.

When the mode is turned on, the following will throw an exception:

```php
$users = User::get();

$users[0]->posts
```

```
StrictLoadingViolationException: Trying to lazy load [posts] in model [User] is restricted
```

However, this will work:

```php
$user = User::find(1);

$user->posts
```

### CODE REVIEW

**Comment by lancepioch** [comment]
Food for thought, it might be better to have this as a default Laravel config setting instead with the ability to override this. Otherwise, it's a simple and clean solution.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -81,6 +81,13 @@ abstract class Model implements Arrayable, ArrayAccess, Jsonable, JsonSerializab
      */
     protected $withCount = [];
 
+    /**
+     * Indicates whether the lazy loading should be prevented.
+     *
+     * @var bool
+     */
+    public $strictLoading;
```
Comment (ahinkle) [inline src/Illuminate/Database/Eloquent/Model.php:89]:
```suggestion
    public $strictLoading = false;
```

Comment (ahinkle) [inline src/Illuminate/Database/Eloquent/Model.php:89]:
```suggestion
    public $strictLoading = false;
```

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 2]
```diff
@@ -426,6 +427,10 @@ protected function getAttributeFromArray($key)
      */
     public function getRelationValue($key)
     {
+        if ($this->strictLoading && ! $this->relationLoaded($key)) {
```
Comment (cerbero90) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php:430]: Another alternative may be moving the condition here and avoid the changes in the Eloquent builder:
```suggestion
        if ($this->getConnection() && $this->getConnection()->getConfig('strict_load') && ! $this->relationLoaded($key)) {
```

**Comment by hkan** [comment]
It might be beneficial to have a configuration option to make it **not** throw an error but trigger an event in the app. So I can log the occurrences or send notifications to myself about it. Wouldn't be as strict, but still... I wouldn't want end users to see error pages just because my code wasn't optimized enough.

**Comment by negoziator** [comment]
I really like this idea.

Would it be possible to add something like Model::enableStrictLoading()
and Model::disableStrictLoading()

Just like with foreignkey checks?

**Comment by mathiasgrimm** [comment]
I think in the first version it could log a deprecation so that people can gradually fix it then on the next version you could throw an exception.

**Comment by mathiasgrimm** [comment]
That would also realize all places that needed to be changed. You might have forgotten some place it dont want to break production

**Pull request closed without merging**